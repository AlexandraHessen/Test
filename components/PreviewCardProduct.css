.PreviewCardProduct{
    display: flex;
    flex-direction: column;
    align-items: center;
    /* row-gap: 1.5rem; */
    width: 25%;
    margin-bottom: 30px;
}
.PreviewCardProduct > a{
    text-align: center;
    text-decoration: none;
    color: #363636;

}
.PreviewCardProduct .NameProduct{
    margin-bottom: 5px;
}


.PreviewCardProduct .PlantImg{
    width: 95%;
    height: 300px;
    margin-bottom: 15px;
}

.PreviewCardProduct .Price{
    font-weight: bold;
    margin-top: 5px;
    display: inline-block;
}

.PreviewCardProduct .fa-shopping-cart{
    color:#357A10;
     border: 1px solid #357A10;
     color: #8ccb68;
     font-size: 17px;
     margin-left: 15px;
     padding: 8px 13px;
    border-radius: 5px;
    border: solid 2px #8ccb68;
     /* position: absolute;
     left: 20px;
     top: 13px; */
}