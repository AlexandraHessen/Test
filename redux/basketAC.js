// ---------------- ActionCreator ----------------//
const ADD_PRODUCT = "ADD_PRODUCT"
const DEL_PRODUCT = "DEL_PRODUCT"
const CLEAR_BASKET = "CLEAR_BASKET" //пользователь сам очищает корзину при желании

//название action type

//функция для dispatch (из файла CounterButton) которая возвращает хэш с action type 
// называется ActionCreator

const add_product = function(productId, objProductInfo) {
    return {
      type: ADD_PRODUCT,
      productId: productId,
      objProductInfo: objProductInfo
    };
  }
  
  const del_product = function(productId) {
    return {
      type: DEL_PRODUCT,
      productId: productId,
    };
  }

  const clear_basket = function() {
    return {
      type: CLEAR_BASKET
    };
  }

export {
    add_product, ADD_PRODUCT,
    del_product, DEL_PRODUCT,
    clear_basket, CLEAR_BASKET,
  }