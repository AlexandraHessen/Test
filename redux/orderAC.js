// ---------------- ActionCreator ----------------//
const ADD_ORDER = "ADD_ORDER"

//название action type

//функция для dispatch (из файла CounterButton) которая возвращает хэш с action type 
// называется ActionCreator

const add_order = function(objCustomerInfo, objOrderInfo) {
    return {
      type: ADD_ORDER,
      objCustomerInfo: objCustomerInfo,
      objOrderInfo: objOrderInfo
    };
  }

export {
    add_order, ADD_ORDER,
  }