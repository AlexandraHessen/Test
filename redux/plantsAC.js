// ---------------- ActionCreator ----------------//
const PLANTS_LOADING='PLANTS_LOADING';
const PLANTS_ERROR='PLANTS_ERROR';
const PLANTS_SET='PLANTS_SET';
//название action type

//функция для dispatch (из файла CounterButton) которая возвращает хэш с action type 
// называется ActionCreator
const plantsLoadingAC=function() {
  return {
    type: PLANTS_LOADING,
  };
}

const plantsErrorAC=function() {
  return {
    type: PLANTS_ERROR,
  };
}

const plantsSetAC=function(plants) {
  return {
    type: PLANTS_SET,
    plants:plants,
  };
}

export {
  plantsLoadingAC,PLANTS_LOADING,
  plantsErrorAC,PLANTS_ERROR,
  plantsSetAC,PLANTS_SET,
}
