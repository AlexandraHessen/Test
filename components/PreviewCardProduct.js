"use strict";

import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import {connect} from 'react-redux';

import {add_product} from '../redux/basketAC'

import "./PreviewCardProduct.css"
class PreviewCardProduct extends React.PureComponent{p

    static propTypes ={
        info: PropTypes.shape({
            code: PropTypes.number.isRequired,
            name: PropTypes.string.isRequired,
            category: PropTypes.string.isRequired,
            price: PropTypes.number.isRequired,
            description: PropTypes.string.isRequired,
            imgUrl: PropTypes.string.isRequired,
        }),
        pageNumber:  PropTypes.number.isRequired
    }

    addProductToBasket = () =>{
        this.props.dispatch( add_product(this.props.info.code,this.props.info) );
        
    }

    render(){
        return (
            <div className="PreviewCardProduct">
                <NavLink to={`/catalog/${this.props.pageNumber}/${this.props.info.code}`} >
                    <img src={this.props.info.imgUrl} className="PlantImg" />
                </NavLink>
                <NavLink to={`/catalog/${this.props.pageNumber}/${this.props.info.code}`} >
                    <p className="NameProduct">{this.props.info.name}</p>
                </NavLink>
                <div>
                    <h4 className="Price">{this.props.info.price} руб.</h4>
                <NavLink to="/basket">
                    <button type="button" onClick = {this.addProductToBasket} className="ButtonToBasket">
                        <i className="fas fa-shopping-cart"></i>
                    </button>
                </NavLink>

                </div>
            </div>
        )
    }
}

const mapStateToProps = function (state) {
    return { }; 
};

export default connect(mapStateToProps)(PreviewCardProduct);
