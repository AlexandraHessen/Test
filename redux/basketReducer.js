// ---------------- САМА ФУНКЦИЯ REDUСER  ----------------//
// REDUCER не делает запрос в Ajax, логика должна быть в компоненте
import { ADD_PRODUCT, DEL_PRODUCT, CLEAR_BASKET } from './basketAC'; //название action type


const initState={ // начальный state
    productsInBasket: {}
    //КЛЮЧИ productId, objProductInfo
    // 1: {code: 1, name: 'Замиокулькас'}

//   status: 0, // 0 - ничего не началось, 1 - идёт загрузка, 2 - была ошибка, 3 - данные загружены
//   data: null, //даные которые мы загрузили по сети

}

function basketReducer(state=initState,action) {
  switch (action.type) {

    case ADD_PRODUCT: {
      console.log('action:',action);
      console.log('state до обработки редьюсером:',state);
      let newState={...state, //старый state
        productsInBasket:{...state.productsInBasket, //заменить старые продукты на теже +
          // ПРИЛЕТИТ какой продукт (productid) изменить и насколько(addvalue)
          [action.productId]:action.objProductInfo //+
          //[action.counterid] - В нужном продукте 
          // state.productsInBasket[action.productid] - его старое значение 
          // action.addvalue - изменить на прилетевшее
        }
      };
      console.log('state после обработки редьюсером:',newState);
      return newState;
    }

    case DEL_PRODUCT: {
      let newState = {...state};
      delete newState.productsInBasket[action.productId];
      return newState;
    }

    case CLEAR_BASKET:{
      let newState = {...state};
      newState.productsInBasket={};
      return newState;
    }

    // case PLANTS_ERROR: {
    //   let newState={
    //     status:2,
    //     data:null,
    //   };
    //   return newState;
    // }

    // case PLANTS_SET: {
    //   let newState={
    //     status:3,
    //     data:action.plants, //загруженные данные
    //   };
    //   return newState;
    // }
    
//обязатнльно пишем на тот случай если пришел action который не относится к этому reducery
    default:
      return state;
  }
}

export default basketReducer;
