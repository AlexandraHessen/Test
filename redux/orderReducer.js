// ---------------- САМА ФУНКЦИЯ REDUСER  ----------------//
// REDUCER не делает запрос в Ajax, логика должна быть в компоненте
import { ADD_ORDER } from './orderAC'; //название action type

const initState={ // начальный state
    orders: []
}

function orderReducer(state=initState,action) {
  switch (action.type) {
    case ADD_ORDER: {
      let newOrder=[...state.orders]
      newOrder.push( { objCustomerInfo: action.objCustomerInfo, products: action.objOrderInfo } );
      let newState={...state, orders: newOrder};

      return newState;
    }

              //обязатнльно пишем на тот случай если пришел action который не относится к этому reducery
    default:
      return state;
  }
}

export default orderReducer;