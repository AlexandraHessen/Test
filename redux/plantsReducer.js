// ---------------- САМА ФУНКЦИЯ REDUСER  ----------------//
// REDUCER не делает запрос в Ajax, логика должна быть в компоненте
import { PLANTS_LOADING, PLANTS_ERROR, PLANTS_SET } from './plantsAC'; //название action type


const initState={ // начальный state

  status: 0, // 0 - ничего не началось, 1 - идёт загрузка, 2 - была ошибка, 3 - данные загружены
  data: null, //даные которые мы загрузили по сети

}

function plantsReducer(state=initState,action) {
  switch (action.type) {

    case PLANTS_LOADING: {
      let newState={
        status:1,
        data:null,
      };
      return newState;
    }

    case PLANTS_ERROR: {
      let newState={
        status:2,
        data:null,
      };
      return newState;
    }

    case PLANTS_SET: {
      let newState={
        status:3,
        data:action.plants, //загруженные данные
      };
      return newState;
    }
    
//обязатнльно пишем на тот случай если пришел action который не относится к этому reducery
    default:
      return state;
  }
}

export default plantsReducer;
