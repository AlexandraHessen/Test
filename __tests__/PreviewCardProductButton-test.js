"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import { createStore } from 'redux';
import combinedReducer from '../redux/reducers.js';
import { Provider } from 'react-redux';
import {add_product} from '../redux/basketAC'

import PreviewCardProduct from '../components/PreviewCardProduct';
                    //импортируем файл который будем тестировать

let info = 	{
    "code": 1,
    "name": "Замиокулькас",
    "category": "deciduous",
    "price": 265,
    "description": "Полив по мере полного просыхания земляного кома, раз в 10-15 дней, в опрыскивании не нуждается. Чрезмерный полив замиокулькасов может приводить к пожелтению листьев.",
    "imgUrl": "/img/img_9241_1.jpg"
}

let pageNumber=1
let store=createStore(combinedReducer);

test('PreviewCardProduct', () => {
                            // создаём тестовую версию компонента который будем тестировать
  const component = renderer.create( //с нужными данными для построения
        <Provider store={store}>
          <BrowserRouter>
            <PreviewCardProduct info = {info} pageNumber = {pageNumber} />  
          </BrowserRouter>
        </Provider>
  );

                            // 1. получаем снэпшот (HTML-снимок) компонента для сверки, чтобы вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

                            // найдём в вёрстке компонента саму кнопку
                            // в верстке <input type="button" value="Все" onClick={this.allClients}/>
                            const buttonAddProductToBasket = component.root.find( el => el.type=='button'); 
                            // нажимаем на кнопку
    buttonAddProductToBasket.props.onClick();

                            // 2. получаем уже изменённый снэпшот
    componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();

    store.dispatch(add_product(info.code,info));
    expect(Object.keys(store.getState().basket.productsInBasket).length).toBe(1); 
});
