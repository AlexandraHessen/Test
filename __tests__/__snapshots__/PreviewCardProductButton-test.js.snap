// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewCardProduct 1`] = `
<div
  className="PreviewCardProduct"
>
  <a
    aria-current={false}
    className={undefined}
    href="/catalog/1/1"
    onClick={[Function]}
    style={undefined}
  >
    <img
      className="PlantImg"
      src="/img/img_9241_1.jpg"
    />
  </a>
  <a
    aria-current={false}
    className={undefined}
    href="/catalog/1/1"
    onClick={[Function]}
    style={undefined}
  >
    <p
      className="NameProduct"
    >
      Замиокулькас
    </p>
  </a>
  <div>
    <h4
      className="Price"
    >
      265
       руб.
    </h4>
    <a
      aria-current={false}
      className={undefined}
      href="/basket"
      onClick={[Function]}
      style={undefined}
    >
      <button
        className="ButtonToBasket"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fas fa-shopping-cart"
        />
      </button>
    </a>
  </div>
</div>
`;

exports[`PreviewCardProduct 2`] = `
<div
  className="PreviewCardProduct"
>
  <a
    aria-current={false}
    className={undefined}
    href="/catalog/1/1"
    onClick={[Function]}
    style={undefined}
  >
    <img
      className="PlantImg"
      src="/img/img_9241_1.jpg"
    />
  </a>
  <a
    aria-current={false}
    className={undefined}
    href="/catalog/1/1"
    onClick={[Function]}
    style={undefined}
  >
    <p
      className="NameProduct"
    >
      Замиокулькас
    </p>
  </a>
  <div>
    <h4
      className="Price"
    >
      265
       руб.
    </h4>
    <a
      aria-current={false}
      className={undefined}
      href="/basket"
      onClick={[Function]}
      style={undefined}
    >
      <button
        className="ButtonToBasket"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fas fa-shopping-cart"
        />
      </button>
    </a>
  </div>
</div>
`;
